openssl/sha.go

```diff
@@ -9,48 +9,122 @@ package openssl
 // #include "goopenssl.h"
 import "C"
 import (
+	"crypto"
 	"errors"
 	"hash"
+	"runtime"
+	"strconv"
 	"unsafe"
 )
 
-// NewSHA1 returns a new SHA1 hash.
-func NewSHA1() hash.Hash {
-	h := new(sha1Hash)
+type evpHash struct {
+	md        *C.EVP_MD
+	ctx       *C.EVP_MD_CTX
+	ctx2      *C.EVP_MD_CTX
+	size      int
+	blockSize int
+}
+
+func newEvpHash(ch crypto.Hash, size, blockSize int) *evpHash {
+	md := cryptoHashToMD(ch)
+	if md == nil {
+		panic("openssl: unsupported hash function: " + strconv.Itoa(int(ch)))
+	}
+	ctx := C.go_openssl_EVP_MD_CTX_new()
+	ctx2 := C.go_openssl_EVP_MD_CTX_new()
+	h := &evpHash{
+		md:        md,
+		ctx:       ctx,
+		ctx2:      ctx2,
+		size:      size,
+		blockSize: blockSize,
+	}
+	runtime.SetFinalizer(h, (*evpHash).finalize)
 	h.Reset()
 	return h
 }
 
-type sha1Hash struct {
-	ctx C.SHA_CTX
-	out [20]byte
+func (h *evpHash) finalize() {
+	C.go_openssl_EVP_MD_CTX_free(h.ctx)
+	C.go_openssl_EVP_MD_CTX_free(h.ctx2)
 }
 
-type sha1Ctx struct {
-	h      [5]uint32
-	nl, nh uint32
-	x      [64]byte
-	nx     uint32
-}
+func (h *evpHash) Reset() {
+	C.go_openssl_EVP_MD_CTX_reset(h.ctx)
```

Comment:
Why doesn't this reset `h.ctx2`?

Reply:
See previous comment for full context. `h.ctx2` is reset every time `Sum` is called.